Update rule APIs for conditional actions (#3320)
elastic/security-docs: docs/detections/api/rules/rules-api-bulk-actions.asciidoc

|==============================================

Optional fields:

[width="100%",options="header"]
|==============================================
|Name |Type |Description

> |frequency |String a|Object containing action’s frequency:

Author: ```suggestion
|frequency |String a|Object containing an action’s frequency:
```

* `summary` (Boolean, required): Defines whether notification will be sent for summary of alerts or for each generated alert individually.

> * `notifyWhen` (String, required`): Defines how often alerts generate actions. Valid values are:

Author: @e40pud is my understanding correct, or should this actually be "rules activated actions" instead? For example:

```suggestion
* `notifyWhen` (String, required`): Defines how often rules activate actions. Valid values are:
```

> * `notifyWhen` (String, required`): Defines how often alerts generate actions. Valid values are:

Reviewer: I think you are right and we can use alerts in this case.

> * `notifyWhen` (String, required`): Defines how often alerts generate actions. Valid values are:

Reviewer 2: Yes, I think your suggestion is right, @nastasha-solomon.

> Defines how often rules run actions. Valid values are...

> ** `onActiveAlert`: Actions run when the alert becomes active and at each check interval while the rule conditions are met.

Author: @e40pud I'm not sure I understand this - when does an alert become active? Is this a concept that's specific to alerts generated by Security detection rules?

Reviewer: I copied description of these values from the alerting framework API docs https://www.elastic.co/guide/en/kibana/current/create-rule-api.html

If I remember correctly, alerting framework addressing security solution rules as alerts. So, in our case we could say `rule becomes active`.

@marshallmain can you confirm that my understanding is correct here?

Reviewer 3: Yeah the alerting framework has a more general description of `onActiveAlert` here that might be confusing to security users. The alerting framework has the capability to track an alert's status across multiple rule executions. E.g. if there's a rule checking for server CPU usage exceeding some threshold, it can keep an alert open and update the alert each time the rule runs and finds that the CPU usage is **still** exceeding the threshold. My understanding is the `at each check interval while the rule conditions are met.` part is intended to describe this behavior where it runs actions for that alert every time the rule runs as long as the CPU usage is still exceeding the threshold.

Alerts for security rules are more "fire and forget" where we don't re-check past alerts on the next rule execution, so actions won't keep firing "while rule conditions are met".

With that in mind, perhaps we can tailor the description here to the more specific security rule implementation and say something like "Actions run when the alert is generated" for `onActiveAlert` and "Actions run on the throttle interval and summarize new alerts from that interval" for `onThrottleInterval`.

Author: ```suggestion
** `onActiveAlert`: Actions run when the alert is generated.
```

> ** `onThrottleInterval`: Actions run when the alert becomes active and at the interval specified in the throttle property while the rule conditions are met.

Author: @e40pud I removed the last part of the description because it seemed more related to what generates the alert, not how the alert is influenced by the `onThrottleInterval` field. Let me know if it's still necessary and I can re-add it in.

```suggestion
** `onThrottleInterval`: Actions run when the alert becomes active and at the interval specified in the throttle property.
```

Reviewer: I copied description of these values from the alerting framework API docs https://www.elastic.co/guide/en/kibana/current/create-rule-api.html

Same as above, we might need to rephrase `alert becomes active` to `rule becomes active`

Author: @marshallmain would you mind reviewing this to make sure it's still technically accurate? TY!

```suggestion
** `onThrottleInterval`: Actions run on the specified throttle interval and summarize new alerts generated during that interval.
```

Reviewer 3: That looks great!

> * `throttle` (String, optional): Defines how often an alert generates repeated actions. This custom action interval must be specified in seconds, minutes, hours, or days. For example, `10m` or `1h`. This property is used only if `notifyWhen` is `onThrottleInterval`.

Author: ```suggestion
* `throttle` (String, optional): Defines how often an alert prompts repeated actions. This custom action interval can be specified in seconds, minutes, hours, or days. For example, `10m` or `1h`. This property is used only if `notifyWhen` is `onThrottleInterval`.
```

Reviewer: Here as well, we might need to use `a rule prompts` instead of `an alert prompts`

Author: Thanks, I'll make this change! I'm also going to make one small edit to the verb because I think "runs" is more appropriate here if we're talking about a rule triggering repeated actions. 

```suggestion
* `throttle` (String, optional): Defines how often rule run repeated actions. This custom action interval must be specified in seconds, minutes, hours, or days. For example, `10m` or `1h`. This property is used only if `notifyWhen` is `onThrottleInterval`.
```

Reviewer: This looks great to me! Thank you!

> |alerts_filter |Object a|Object containing action’s conditional filters:

Author: ```suggestion
|alerts_filter |Object a|Object containing an action’s conditional filters:
```

> * `timeframe` (Object, optional): Object containing the time frame during which this action can be fired.

Author: ```suggestion
* `timeframe` (Object, optional): Object containing the time frame for when this action can be activated.
```

> ** `days` (Array of integers, required): List of days of the week on which this action can fire, expressed as numbers between `1-7`, where `1` is Monday and `7` is Sunday. An empty array is equivalent to all days of the week being selected.

Author: ```suggestion
** `days` (Array of integers, required): List of days of the week on which this action can activate. Days of the week are expressed as numbers between `1-7`, where `1` is Monday and `7` is Sunday. To select all days of the week, enter an empty array.
```

> ** `hours` (Object, required): The hours of the day during which this action can fire. Expressed as two strings in the format `hh:mm` in `24` hour time. A start of `00:00` and an end of `24:00` means the action can fire all day.

Author: ```suggestion
** `hours` (Object, required): The hours of the day during which this action can activate. Hours of the day are expressed as two strings in the format `hh:mm` in `24` hour time. A start of `00:00` and an end of `24:00` means the action can activate all day.
```

*** `start` (String, required)
*** `end` (String, required)

** `timezone` (String, required): An ISO timezone name, such as `Europe/Madrid` or `America/New_York`. Specific offsets such as UTC or UTC+1 will also work, but lack built-in DST.

> * `query` (Object, optional): Object containing a query filter to apply to the action to determine whether or not it should fire.

Author: ```suggestion
* `query` (Object, optional): Object containing a query filter which gets applied to an action and determines whether the action should activate. 
```

** `kql` (String, required): A KQL string.
** `filters` (Array of objects, required): A filter object, as defined in the `kbn-es-query` package.

|==============================================

[discrete]
[[throttle-schema-bulk]]
===== `throttle` schema

`throttle` defines the maximum interval in which a rule's actions are executed. It accepts the following values:

> **NOTE**: In {elastic-sec} 8.8 and later, we don't recommend using this field as actions are now set on an individual action level using `frequency` field`.  If some of the actions don’t have frequency attribute set then the `throttle` field will be converted into an action level frequency and set to any action without a frequency.

Author: Note to self: grabbed this from the google doc. Feedback is pending.

Author: Possible revision:

```suggestion
**NOTE**: In {elastic-sec} 8.8 and later, the `throttle` setting is only applied to actions without action-level `frequency` settings. In the response for these types of actions, the `throttle` setting is moved into the throttle  and notifyWhen fields within individual actions’ frequency field.

```

** `hours` (Object, required): The hours of the day during which this action can run. Hours of the day are expressed as two strings in the format `hh:mm` in `24` hour time. A start of `00:00` and an end of `24:00` means the action can run all day.
*** `start` (String, required)
*** `end` (String, required)

** `timezone` (String, required): An ISO timezone name, such as `Europe/Madrid` or `America/New_York`. Specific offsets such as UTC or UTC+1 will also work, but lack built-in DST.

* `query` (Object, optional): Object containing a query filter which gets applied to an action and determines whether the action should run. 
** `kql` (String, required): A KQL string.
** `filters` (Array of objects, required): A filter object, as defined in the `kbn-es-query` package.

|==============================================

[discrete]
[[throttle-schema-bulk]]
===== `throttle` schema

`throttle` defines the maximum interval in which a rule's actions are executed. It accepts the following values:

> **NOTE**: In {elastic-sec} 8.8 and later, the rule level `throttle` setting is applied to any actions missing action level frequency settings. In the response, the updated action frequencies are shown in the `frequency` field for individual actions. For more information about action level frequency settings, refer <<optional-actions-fields-bulk-update>>.

Author: ```suggestion
**NOTE**: In {elastic-sec} 8.8 and later, you can use the (<<optional-actions-fields-bulk-update,`frequency`>>) field to define frequencies for individual actions. Actions without frequencies will acquire a converted version of the rule's `throttle` field. In the response, the converted `throttle` setting appears in the individual actions' `frequency` field.
```

> **NOTE**: In {elastic-sec} 8.8 and later, you can use the (<<optional-actions-fields-bulk-update,`frequency`>>) field to define frequencies for individual actions. Actions without frequencies will acquire a converted version of the rule's `throttle` field. In the response, the converted `throttle` setting appears in the individual actions' `frequency` field.

Reviewer 2: I think the wording is ok. But why is it different from the notes in other parts of this PR like [here](https://security-docs_3320.docs-preview.app.elstc.co/guide/en/security/master/rules-api-create.html#opt-fields-all)?

> **NOTE**: In {elastic-sec} 8.8 and later, you can use the (<<optional-actions-fields-bulk-update,`frequency`>>) field to define frequencies for individual actions. Actions without frequencies will acquire a converted version of the rule's `throttle` field. In the response, the converted `throttle` setting appears in the individual actions' `frequency` field.

Author: @peluja1012 those are outdated versions of the note. I wanted to get this version reviewed before applying the changes elsewhere. I'll update the outdated notes now so everything matches. Thanks for the review!

|==============================================

[discrete]
[[optional-actions-fields-bulk-update]]
===== Optional `action` fields

[width="100%",options="header"]
|==============================================
|Name |Type |Description

|frequency |String a|Object containing an action’s frequency:

> * `summary` (Boolean, required): Defines whether to send notifications as a summary of alerts or for each generated alert.

Author: So your edit is partially accurate, but I don't want to remove "notifications" because this field dictates how notifications are sent (i.e., as alert summaries or as individual notices).

```suggestion
* `summary` (Boolean, required): Defines whether to send notifications alert summaries or for individual alerts. 
```

* `notifyWhen` (String, required`): Defines how often rules run actions. Valid values are:

> ** `onActiveAlert`: Actions run when the alert is generated.

Reviewer 4: ```suggestion
** `onActiveAlert`: Actions run when an alert is generated.
```

** `onThrottleInterval`: Actions run on the specified throttle interval and summarize new alerts generated during that interval.

* `throttle` (String, optional): Defines how often a rule runs reoccurring actions. This custom action interval must be specified in seconds, minutes, hours, or days. For example, `10m` or `1h`. This property is used only if `notifyWhen` is `onThrottleInterval`.

|alerts_filter |Object a|Object containing an action’s conditional filters:

> * `timeframe` (Object, optional): Object containing the time frame for when this action can be run.

Reviewer 4: ```suggestion
* `timeframe` (Object, optional): Defines the time frame when this action can run.
```
Already said it's an object

Author: Good catch on the duplicate details

> ** `days` (Array of integers, required): List of days of the week on which this action can be run. Days of the week are expressed as numbers between `1-7`, where `1` is Monday and `7` is Sunday. To select all days of the week, enter an empty array.

Reviewer 4: ```suggestion
** `days` (Array of integers, required): Days of the week when this action can run, expressed as numbers between `1-7`, where `1` is Monday and `7` is Sunday. To select all days of the week, enter an empty array.
```

Author: Minor edit:

```suggestion
** `days` (Array of integers, required): Days of the week when this action will run, expressed as numbers between `1-7`, where `1` is Monday and `7` is Sunday. To select all days of the week, enter an empty array.
```

> ** `hours` (Object, required): The hours of the day during which this action can run. Hours of the day are expressed as two strings in the format `hh:mm` in `24` hour time. A start of `00:00` and an end of `24:00` means the action can run all day.

Reviewer 4: ```suggestion
** `hours` (Object, required): Hours of the day during which this action can run, expressed as two strings in the format `hh:mm` in 24 hour time. A start of `00:00` and an end of `24:00` means the action can run all day.
```

> ** `hours` (Object, required): The hours of the day during which this action can run. Hours of the day are expressed as two strings in the format `hh:mm` in `24` hour time. A start of `00:00` and an end of `24:00` means the action can run all day.

Reviewer 2: I do not. @e40pud do you know if the code has it as `<=  24:00` or `< 24:00`?

> ** `hours` (Object, required): The hours of the day during which this action can run. Hours of the day are expressed as two strings in the format `hh:mm` in `24` hour time. A start of `00:00` and an end of `24:00` means the action can run all day.

Reviewer 5: `24:00` is a special case in the code, it's interpreted as turning off the `hours` filter entirely. We recommend `24:00` for clarity, confirming with @ersin-erdal but if I remember correctly, I believe setting `start` and `end` both to `00:00` does the same thing.

> ** `hours` (Object, required): The hours of the day during which this action can run. Hours of the day are expressed as two strings in the format `hh:mm` in `24` hour time. A start of `00:00` and an end of `24:00` means the action can run all day.

Reviewer 6: Yes, `00:00 -> 00:00` and `00:00 -> 24:00` are the same thing.
But the component we used on UI does not allow `00:00` as end time, it replaces `24:00` with `00:00`
But form the API perspective 24:00 is still usable.
Besides that, as long as the end time is lower than the start time or equals to it, the time range is considered overnight, which means full 24 hours.
So, `15:00 -> 15:00` or `21:33 -> 21:33` are also the same thing.

> ** `hours` (Object, required): The hours of the day during which this action can run. Hours of the day are expressed as two strings in the format `hh:mm` in `24` hour time. A start of `00:00` and an end of `24:00` means the action can run all day.

Author: Thanks for elaborating on this, @ersin-erdal! I'm wondering if we should add this part, or if it's well-known enough that it wouldn't be worth mentionin:

> Besides that, as long as the end time is earlier than the start time or equals to it, the time range is considered overnight, which means full 24 hours.
> So, 15:00 -> 15:00 or 21:33 -> 21:33 are also the same thing.

@e40pud @joepeeples you think this info would be helpful to include?

> ** `hours` (Object, required): The hours of the day during which this action can run. Hours of the day are expressed as two strings in the format `hh:mm` in `24` hour time. A start of `00:00` and an end of `24:00` means the action can run all day.

Reviewer 7: Might be worth mentioning -- I don't know what's common knowledge for API programmers but I wouldn't have personally guessed that's how time values work.

*** `start` (String, required)
*** `end` (String, required)

> ** `timezone` (String, required): An ISO timezone name, such as `Europe/Madrid` or `America/New_York`. Specific offsets such as UTC or UTC+1 will also work, but lack built-in DST.

Reviewer 5: Yes these are exact values. Also, just realized that `ISO timezone name` is inaccurate, this should actually be `IANA timezone name` (https://en.wikipedia.org/wiki/Tz_database)

> ** `timezone` (String, required): An ISO timezone name, such as `Europe/Madrid` or `America/New_York`. Specific offsets such as UTC or UTC+1 will also work, but lack built-in DST.

Author: Leaving this for posterity: @Zacqary fixed this in https://github.com/elastic/security-docs/pull/3373. 

Thanks, all!

> * `query` (Object, optional): Object containing a query filter that gets applied to an action and determines whether the action should run. 

Reviewer 4: ```suggestion
* `query` (Object, optional): Object containing a query filter that gets applied to an action to determine whether the action should run. 
```

> * `query` (Object, optional): Object containing a query filter which gets applied to an action and determines whether the action should run. 

Reviewer 7: ```suggestion
* `query` (Object, optional): Object containing a query filter that gets applied to an action and determines whether the action should run. 
```

** `kql` (String, required): A KQL string.
** `filters` (Array of objects, required): A filter object, as defined in the `kbn-es-query` package.

|==============================================

[discrete]
[[throttle-schema-bulk]]
===== `throttle` schema

`throttle` defines the maximum interval in which a rule's actions are executed. It accepts the following values:

[NOTE]

=========

> The rule level `throttle` field is deprecated in {elastic-sec} 8.8 and is scheduled for end of life in Q4.

Author: ```suggestion
The rule level `throttle` field is deprecated in {elastic-sec} 8.8 and is scheduled for end of life in Q4 of 2024.
```

> The rule level `throttle` field is deprecated in {elastic-sec} 8.8 and is scheduled for end of life in Q4 of 2024.

Reviewer 7: ```suggestion
The rule-level `throttle` field is deprecated in {elastic-sec} 8.8 and is scheduled for end of life in Q4 of 2024.
```

> The rule level `throttle` field is deprecated in {elastic-sec} 8.8 and will remain active for at least the next 12 months.

Reviewer 2: Hi @nastasha-solomon, could we please update this text to use language similar to previous deprecations like this?
> The rule level `throttle` field is deprecated in {elastic-sec} 8.8 and is scheduled for end of life in Q4 2024.

Author: ```suggestion
The rule level `throttle` field is deprecated in {elastic-sec} 8.8 and is scheduled for end of life in Q4.
```

In {elastic-sec} 8.8 and later, you can use the (<<optional-actions-fields-bulk-update,`frequency`>>) field to define frequencies for individual actions. Actions without frequencies will acquire a converted version of the rule's `throttle` field. In the response, the converted `throttle` setting appears in the individual actions' `frequency` field.

> =========

Author: @peluja1012 @e40pud the first sentence in this note (line 447) is the deprecation notice. Let me know it's correct or if any other changes are needed. Thank you!
